lib/utils/object.ts

```diff
@@ -0,0 +1,172 @@
+    // In any case, don't clone stuff that's an object, but not a plain one - fx example sequelize models and instances
+    if (onlyPlain || typeof elem === 'object') {
+      return elem;
+    }
+
+    // Preserve special data-types like `fn` across clones. _.get() is used for checking up the prototype chain
+    if (elem && typeof elem.clone === 'function') {
+      return elem.clone();
+    }
+  });
+}
+
+/**
+ * Receives a tree-like object and returns a plain object which depth is 1.
+ *
+ * - Input:
+ *
+ *  {
+ *    name: 'John',
+ *    address: {
+ *      street: 'Fake St. 123',
+ *      coordinates: {
+ *        longitude: 55.6779627,
+ *        latitude: 12.5964313
+ *      }
+ *    }
+ *  }
+ *
+ * - Output:
+ *
+ *  {
+ *    name: 'John',
+ *    address.street: 'Fake St. 123',
+ *    address.coordinates.latitude: 55.6779627,
+ *    address.coordinates.longitude: 12.5964313
+ *  }
+ *
+ * @param {object} value an Object
+ * @returns {object} a flattened object
+ * @private
+ */
+export function flattenObjectDeep(value: object) {
+  if (!_.isPlainObject(value)) return value;
+  const flattenedObj: { [key: string]: any } = {};
+
+  function flattenObject(obj: { [key: string]: any }, subPath?: string) {
+    Object.keys(obj).forEach(key => {
+      const pathToProperty = subPath ? `${subPath}.${key}` : key;
+      if (typeof obj[key] === 'object' && obj[key] !== null) {
+        flattenObject(obj[key], pathToProperty);
+      } else {
+        flattenedObj[pathToProperty] = _.get(obj, key);
+      }
+    });
+    return flattenedObj;
+  }
+
+  return flattenObject(value, undefined);
+}
+
+/**
+ * Assigns own and inherited enumerable string and symbol keyed properties of source
+ * objects to the destination object.
+ *
+ * https://lodash.com/docs/4.17.4#defaults
+ *
+ * **Note:** This method mutates `object`.
+ *
+ * @param {object} object The destination object.
+ * @param {...object} [sources] The source objects.
+ * @returns {object} Returns `object`.
+ * @private
+ */
+export function defaults(
+  object: { [key: string | symbol]: any },
+  ...sources: { [key: string | symbol]: any }[]
+): object {
+  object = Object(object);
+
+  sources.forEach(source => {
+    if (source) {
+      source = Object(source);
+
+      getComplexKeys(source).forEach((key: string | symbol) => {
+        const value = object[key];
+        const objectPrototype: { [key: string | symbol]: any } =
+          Object.prototype;
+
+        if (
+          value === undefined ||
+          _.eq(value, objectPrototype[key]) &&
+            !objectPrototype.hasOwnProperty.call(object, key)
+        ) {
+          object[key] = source[key];
+        }
+      });
+    }
+  });
+
+  return object;
+}
+
+/**
+ * Returns an new Object which keys are camelized
+ *
+ * @param {object} obj
+ * @returns {string}
+ * @private
+ */
+export function camelizeObjectKeys(obj: { [key: string]: any }) {
```

Comment:
I got carried away and started work on typing that changes the keys to make them camelcased.

It's not finished but all that's missing is listing all the whitespace characters matched by \s & trim.

I can continue working on this later

```typescript
type CamelCaseImpl<S extends string, Separator extends string> =
  S extends `${infer P1}${Separator}${infer P2}${infer P3}`
    ? CamelCaseImpl<`${P1}${Uppercase<P2>}${P3}`, Separator>
  : S;

// must be nested, cannot be CamelCaseImpl<S, '-' | '_' | ' '>
//  or strings like 'first-second_third' will not parse properly.
type Camelize<S extends string> = CamelCaseImpl<CamelCaseImpl<CamelCaseImpl<Trim<S>, '-'>, '_'>, ' '>;

// according to https://tc39.es/ecma262/multipage/text-processing.html#sec-string.prototype.trim
//                <SPACE> <TAB>   <ZWNBSP>
type WhiteSpace = ' ' | '\t' | '\u2060';
// <VT>
// <FF>
// <USP>; // https://www.compart.com/en/unicode/category/Zs

//                    <LF>   <CR>
type LineTerminator = '\n' | '\r';
// <LS>
// <PS>

type Trimable = WhiteSpace | LineTerminator;

type TrimRight<T extends string> = T extends `${infer R}${Trimable}` ? TrimRight<R> : T;
type TrimLeft<T extends string> = T extends `${Trimable}${infer R}` ? TrimLeft<R> : T;
type Trim<T extends string> = TrimRight<TrimLeft<T>>;

const a = {
  test_test: 5,
  '  first--second_third FOURTH   --  test  \t': 10,
};

function camelizeObjectKeys<In extends object>(input: In): {
  [key in keyof In as key extends string ? Camelize<key> : key]: In[key]
} {
  return input;
}

const b = camelizeObjectKeys(a);
// b is {testTest: number, firstSecondThirdFOURTHTest: number}
```

Reply:
Guess this comment can be resolved right?

Reply:
I haven't added it yet because it's not quite right but I'll back it up to a gist and add it later if it ends up being useful